File: go/binary-search/binary_search.go
package binarysearch

func SearchInts(data []int, search int) int {

	lo, hi := 0, len(data)-1
	for lo <= hi {
		mid := lo + (hi-lo)/2
		d := data[mid]
		switch {
		case d < search:
			lo = mid + 1
		case d > search:
			hi = mid - 1
		default:
			return mid
		}
	}
	return -1
}

File: elixir/binary-search/lib/binary_search.ex
defmodule BinarySearch do
  def search(numbers, target) do
    aux(numbers, 0, tuple_size(numbers) - 1, target)
  end
  def aux(_, lo, hi, _) when lo > hi, do: :not_found
  def aux(numbers, lo, hi, target) do
    mid = div(lo + hi, 2)
    cond do
      elem(numbers, mid) == target ->
        {:ok, mid}
      elem(numbers, mid) > target ->
        aux(numbers, lo, mid - 1, target)
      true ->
        aux(numbers, mid + 1, hi, target)
    end
  end
end
